src/Features/LanguageServer/Protocol/Features/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.CompilationManager.cs

```diff
@@ -27,12 +27,19 @@ internal partial class DiagnosticIncrementalAnalyzer
 
             var ideOptions = AnalyzerService.GlobalOptions.GetIdeAnalyzerOptions(project);
 
-            if (_projectCompilationsWithAnalyzers.TryGetValue(project, out var compilationWithAnalyzers) &&
-                ((WorkspaceAnalyzerOptions)compilationWithAnalyzers!.AnalysisOptions.Options!).IdeOptions == ideOptions)
+            if (_projectCompilationsWithAnalyzers.TryGetValue(project, out var compilationWithAnalyzers))
             {
-                // we have cached one, return that.
-                AssertAnalyzers(compilationWithAnalyzers, stateSets);
-                return compilationWithAnalyzers;
+                // We may have cached a null entry if we determiend that there are no actual analyzers to run.
+                if (compilationWithAnalyzers is null)
+                {
+                    return null;
+                }
+                else if (((WorkspaceAnalyzerOptions)compilationWithAnalyzers.AnalysisOptions.Options!).IdeOptions == ideOptions)
```

Comment:
@tmat Something still seems wrong here so if this is different, we'll fall back and call CreateCompilationWithAnalyzersAsync again. But then we still try to add it to the CWT, will see there's already an old value in the key, and end up using the old value. Is this just broken?

Reply:
That does seem broken

Reply:
Going to file a new bug for that.

Reply:
Filed https://github.com/dotnet/roslyn/issues/62550.